File: java/src/main/java/com/quant/localvol/Market.java
package com.quant.localvol;

/**
 * Market description shared by all pricers.
 *
 * <p>A single record covers both asset classes:
 * <ul>
 *   <li><b>Equity</b>: {@code rate} is the risk-free rate {@code r} and
 *       {@code dividend} is the continuous dividend yield {@code q}.</li>
 *   <li><b>FX (Garman-Kohlhagen)</b>: {@code rate} is the domestic rate
 *       {@code rd} and {@code dividend} plays the role of the foreign rate
 *       {@code rf}. Under Garman-Kohlhagen the foreign rate enters every
 *       formula exactly where the dividend yield enters the equity formula,
 *       so one parametrisation serves both; use {@link #fx} to make the
 *       intent explicit at call sites.</li>
 * </ul>
 * The forward is {@code F(T) = S * exp((rate - dividend) * T)} in both cases.
 *
 * @param spot     spot price {@code S0} (equity price or FX rate); strictly positive
 * @param rate     continuously-compounded discount rate {@code r} ({@code rd} for FX); may be negative
 * @param dividend continuous dividend yield {@code q} ({@code rf} for FX); may be negative
 */
public record Market(double spot, double rate, double dividend) {

    /** Validates finiteness of every field and positivity of the spot. */
    public Market {
        if (!Double.isFinite(spot)) {
            throw new IllegalArgumentException("Market.spot must be finite, got " + spot);
        }
        if (!Double.isFinite(rate)) {
            throw new IllegalArgumentException("Market.rate must be finite, got " + rate);
        }
        if (!Double.isFinite(dividend)) {
            throw new IllegalArgumentException("Market.dividend must be finite, got " + dividend);
        }
        if (spot <= 0.0) {
            throw new IllegalArgumentException("Market.spot must be > 0, got " + spot);
        }
    }

    /**
     * Garman-Kohlhagen market: domestic rate {@code rd}, foreign rate {@code rf}.
     *
     * @param spot FX spot (domestic units per foreign unit)
     * @param rd   domestic rate
     * @param rf   foreign rate
     * @return the equivalent {@code Market}
     */
    public static Market fx(double spot, double rd, double rf) {
        return new Market(spot, rd, rf);
    }

    /**
     * Forward {@code F(T) = S0 * exp((r - q) T)}.
     *
     * @param expiry maturity in years, finite and {@code >= 0}
     * @return the forward level
     */
    public double forward(double expiry) {
        if (!Double.isFinite(expiry) || expiry < 0.0) {
            throw new IllegalArgumentException("expiry must be finite and >= 0, got " + expiry);
        }
        return spot * Math.exp((rate - dividend) * expiry);
    }

    /**
     * {@code ln F(T)} — convenient for log-moneyness lookups.
     *
     * @param expiry maturity in years, finite and {@code >= 0}
     * @return the log-forward
     */
    public double logForward(double expiry) {
        if (!Double.isFinite(expiry) || expiry < 0.0) {
            throw new IllegalArgumentException("expiry must be finite and >= 0, got " + expiry);
        }
        return Math.log(spot) + (rate - dividend) * expiry;
    }
}

File: python/src/localvol/market.py
"""Market description shared by all pricers.

A single dataclass covers both asset classes:

* **Equity**: ``rate`` is the risk-free rate ``r`` and ``dividend`` is the
  continuous dividend yield ``q``.
* **FX (Garman-Kohlhagen)**: ``rate`` is the domestic rate ``rd`` and
  ``dividend`` plays the role of the foreign rate ``rf``.  Under
  Garman-Kohlhagen the foreign rate enters every formula exactly where the
  dividend yield enters the equity formula, so one parametrisation serves
  both; use :meth:`Market.fx` to make the intent explicit at call sites.

The forward is ``F(T) = S * exp((rate - dividend) * T)`` in both cases.
"""

from __future__ import annotations

import math
from dataclasses import dataclass


@dataclass(frozen=True)
class Market:
    """Spot / rates container.

    Attributes:
        spot: Spot price ``S0`` (equity price or FX rate, domestic per foreign).
            Must be strictly positive.
        rate: Continuously-compounded discount rate ``r`` (domestic rate ``rd``
            for FX).  May be negative (EUR/CHF style regimes are supported).
        dividend: Continuous dividend yield ``q`` (foreign rate ``rf`` for FX).
            May be negative.
    """

    spot: float
    rate: float
    dividend: float = 0.0

    def __post_init__(self) -> None:
        for name in ("spot", "rate", "dividend"):
            v = getattr(self, name)
            if not math.isfinite(v):
                raise ValueError(f"Market.{name} must be finite, got {v!r}")
        if self.spot <= 0.0:
            raise ValueError(f"Market.spot must be > 0, got {self.spot!r}")

    @classmethod
    def fx(cls, spot: float, rd: float, rf: float) -> "Market":
        """Garman-Kohlhagen market: domestic rate ``rd``, foreign rate ``rf``."""
        return cls(spot=spot, rate=rd, dividend=rf)

    def forward(self, expiry: float) -> float:
        """Forward ``F(T) = S0 * exp((r - q) T)``.  Requires ``expiry >= 0``."""
        if not math.isfinite(expiry) or expiry < 0.0:
            raise ValueError(f"expiry must be finite and >= 0, got {expiry!r}")
        return self.spot * math.exp((self.rate - self.dividend) * expiry)

    def log_forward(self, expiry: float) -> float:
        """``ln F(T)`` — convenient for log-moneyness lookups."""
        return math.log(self.spot) + (self.rate - self.dividend) * expiry
